### PR TITLE
Fix Task can stay running forever

### PR DESCRIPTION
**Pull request by fviale**
If a task launcher dies because the underlying node is cleaned. The task will be seen by the scheduler as running forever.

 Even if the fact that the task launcher dies is not normal, there should be a configurable timeout to control the scheduler behavior with regard to this task.

 This change is to implement a timeout for task launcher failure.

### CODE REVIEW

File: scheduler/scheduler-server/src/test/java/functionaltests/job/error/TestTaskRestartOnNodeFailure.java [hunk 1]
```diff
@@ -151,6 +152,48 @@ private void testTaskKillNode(FileLock fileLock, boolean waitBeforeKill) throws
...
+
+        log("Wait when task starts");
+        schedulerHelper.waitForEventTaskRunning(jobId, "Test task");
+
+        /*
+         * Wait some time to make sure the TaskLauncher is deployed
+         */
+
+        log("Wait some time");
+        Thread.sleep(5000);
```
Comment (sonarqube-activeeon) [inline scheduler/scheduler-server/src/test/java/functionaltests/job/error/TestTaskRestartOnNodeFailure.java]: MAJOR Remove this use of "Thread.sleep()". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2925)

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 7 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 6 minor

Watch the comments in this conversation to review them.

#### 6 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [SchedulingService.java#L59](https://github.com/ow2-proactive/scheduling/blob/95f738aa1aa7b11dce456f28c77b29ee7d05bc18/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L59): Remove this unused import 'org.ow2.proactive.scheduler.common.task.TaskInfo'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AUselessImportCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L633](https://github.com/ow2-proactive/scheduling/blob/95f738aa1aa7b11dce456f28c77b29ee7d05bc18/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L633): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L665](https://github.com/ow2-proactive/scheduling/blob/95f738aa1aa7b11dce456f28c77b29ee7d05bc18/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L665): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L697](https://github.com/ow2-proactive/scheduling/blob/95f738aa1aa7b11dce456f28c77b29ee7d05bc18/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L697): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L727](https://github.com/ow2-proactive/scheduling/blob/95f738aa1aa7b11dce456f28c77b29ee7d05bc18/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L727): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)
1. ![MINOR][MINOR] [SchedulingService.java#L755](https://github.com/ow2-proactive/scheduling/blob/95f738aa1aa7b11dce456f28c77b29ee7d05bc18/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java#L755): Remove the redundant '!unknownSymbol!' thrown exception declaration(s). [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/RunningTaskData.java [hunk 2]
```diff
@@ -44,7 +44,11 @@
 
     private final NodeSet nodes;
 
-    private int pingAttempts = 0;
+    private volatile int pingAttempts = 0;
+
+    private volatile boolean isRestarting = false;
```
Comment (Mykhailenko) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/RunningTaskData.java]: I do not insist but I thought it is kinda bad practive to start name of boolean with "is".

Comment (fviale) [reply scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/RunningTaskData.java]: ok, I can rename the variable.

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java [hunk 3]
```diff
@@ -1049,24 +1050,34 @@ private void jobsRecovered(Collection<InternalJob> jobs) {
...
                 //if progress != previously set progress (0 by default) -> update
                 listener.taskStateUpdated(taskData.getUser(),
-                                          new NotificationData<TaskInfo>(SchedulerEvent.TASK_PROGRESS,
-                                                                         new TaskInfoImpl((TaskInfoImpl) task.getTaskInfo())));
+                                          new NotificationData<>(SchedulerEvent.TASK_PROGRESS,
+                                                                 new TaskInfoImpl((TaskInfoImpl) task.getTaskInfo())));
             }
         } catch (Throwable t) {
-            tlogger.debug(task.getId(), "TaskLauncher is not accessible, checking if the node can be reached.", t);
+            if (taskData.getFirstTaskLauncherFailureTime() < 0) {
```
Comment (Mykhailenko) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java]: maybe you could add method to RunningTaskData, like 'taskLauncherNeverFailed()
'

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java [hunk 4]
```diff
@@ -1049,24 +1050,34 @@ private void jobsRecovered(Collection<InternalJob> jobs) {
...
+                taskData.setFirstTaskLauncherFailureTime(System.currentTimeMillis());
```
Comment (Mykhailenko) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java]: also we can add method: RunningTaskData::firstFailure()

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java [hunk 5]
```diff
@@ -1078,17 +1089,29 @@ private void pingTaskNodeAndInitiateRestart(InternalTask task) {
...
+                    runningTask.setRestarting(true);
+                    restartTaskOnNodeFailure(task);
+                }
+
             } catch (Exception e) {
                 int attempts = runningTask.increaseAndGetPingAttempts();
                 String nodeUrl = nodeUsedToExecuteTask.getNodeInformation().getURL();
                 if (attempts > PASchedulerProperties.SCHEDULER_NODE_PING_ATTEMPTS.getValueAsInt()) {
                     tlogger.error(task.getId(), "node failed " + nodeUrl + ", initiate task restart.", e);
+                    runningTask.setRestarting(true);
```
Comment (Mykhailenko) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java]: You use setRestarting alsways with true argument

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/RunningTaskData.java [hunk 6]
```diff
@@ -44,7 +44,11 @@
 
     private final NodeSet nodes;
 
-    private int pingAttempts = 0;
+    private volatile int pingAttempts = 0;
```
Comment (Mykhailenko) [inline scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/RunningTaskData.java]: According to this: https://stackoverflow.com/questions/7805192/is-a-volatile-int-in-java-thread-safe, volatile int won't be thread safe if you do increment. so maybe it is better to add **sycnhronized**  to method where `pingAttempts++`

Comment (marcocast) [reply scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/RunningTaskData.java]: AtomicInteger ?

Comment (Mykhailenko) [reply scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/RunningTaskData.java]: AtomicInteger is even better

File: scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java (hunk 4 above)
Comment (fviale) [reply scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingService.java]: ok, I will refactor a bit the new methods in RunningTaskData

**Comment by fviale** [comment]
After several discussions, it turns out this change is not appropriate as it will hide errors instead of solving them.

**Pull request closed without merging**